devicetree: default absent label to empty string

pabigot opened the pull request:
When devices are defined using devicetree nodes, the `label` property is used as the driver name.  As we move to making the `label` property optional, we need to account for widespread use of `dev->name` in contexts where the pointer is not validated before it is dereferenced.

Explicitly recognize both a null name and an empty name as referring to no device, and use an empty name as the device name for devices defined from a devicertee node that has no label.

pabigot committed:
device: binding lookup should return null for unsupported names

A null device name should map to a null device.  So should a name that
is empty.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Commits #32141, #32171 through #32175, #32214, #32217, #32219, #32227, and #32242 through #32251 all remove label properties, so it could be good to get this or another solution in before they go in.

(I may not have that list completely correct; not all the clock control ones might change `label`, but most/all the pinmux ones do.)

galak reviewed (changes requested):
I'd rather we leave the DT_LABEL() as is and tweak the DT_PROP_OR() to return "" instead of NULL.

pabigot committed:
device: use empty string as drvname when devicetree label is absent

As we move to making the `label` property optional, we need to account
for widespread use of `dev->name` in contexts where the pointer is not
validated before it is dereferenced.  Use an empty string instead of a
null pointer when no label is provided.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
> I'd rather we leave the DT_LABEL() as is and tweak the DT_PROP_OR() to return "" instead of NULL.

Yes, that seems to be safer as it produces a build error if code tries to use `DT_LABEL()` when that's missing.  Updated.
